File: src/Data_Structure/Tree/Leetcode_Problems/Distribute_coins_in_binary_tree/distributeCoins.cpp
#include <bits/stdc++.h>
using namespace std;

// Binary Tree Node
struct TreeNode {
	int val;
	struct TreeNode *left, *right;
};
// function to create a new node
TreeNode* newNode(int val)
{
	TreeNode* temp = new TreeNode;
	temp->val = val;
	temp->left = temp->right = NULL;
	return (temp);
}
int distributeCoinsUtil(TreeNode* root, int& ans)
{
	// Base Case
	if (root == NULL)
		return 0;
	// Traverse left subtree
	int l = distributeCoinsUtil(root->left, ans);
	// Traverse right subtree
	int r = distributeCoinsUtil(root->right, ans);
	// Update number of moves
	ans += abs(l) + abs(r);
	// Return number of moves to balance current node
	return root->val + l + r - 1;
}
// Function to find the number of moves to distribute all of the coins
int distributeCoins(TreeNode* root)
{
	int ans = 0;
	distributeCoinsUtil(root, ans);
	return ans;
}
int main()
{
	TreeNode* root = newNode(3);
	root->left = newNode(0);
	root->right = newNode(0);
	cout << distributeCoins(root);
	return 0;
}

File: src/Data_Structure/Tree/Leetcode_Problems/Distribute_coins_in_binary_tree/DistributeCoins.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int distributeCoins(TreeNode root) {
        int res = 0;
        if (root.left != null) {
            // Traverse left subtree
            res += distributeCoins(root.left);
            root.val += root.left.val - 1;
            // Update number of moves
            res += Math.abs(root.left.val - 1);
        }
        if (root.right != null) {
            // Traverse right subtree
            res += distributeCoins(root.right);
            root.val += root.right.val - 1;
            // Update number of moves
            res += Math.abs(root.right.val - 1);
        }
        return res;
    }
}